feat: add support to use encoders

DeboraTahara opened the pull request:
# What is this PR for
[sc-74445]

# How I did it
<!-- Explain a little bit of your implementation -->

# Results, How I tested
<!-- Explain how you tested you PR, if there is a visual output,
     a GIF or image is welcome -->

# Checklist
- [x] Assign yourself  in the PR
- [ ] Write unit tests (when relevant)
- [ ] Run rubocop and rake tests locally
- [ ] If this PR initialize a CMAKE package please [enable styling and linting](https://github.com/tidewise/wetpaint-buildconf/blob/master/README.md#enabling-styling-and-linting-checks-for-a-cmake-package)
- [ ] Analyse if this PR modifies the UI, if so:
    - [ ] Tested Tupan's simulation (Charts) :world_map:
    - [ ] Tested Tupan's simulation (Hud) :joystick:
    - [ ] Tested ROV's simulation :joystick:

DeboraTahara committed:
feat: Define CANOPEN object for EncoderCounter

doudou commented on src/DriverBase.cpp:
```diff
@@ -215,6 +263,15 @@ int DriverBase::setupAnalogTPDOs(std::vector<canbus::Message>& messages,
 
+int DriverBase::setupEncoderTPDOs(std::vector<canbus::Message>& messages,
+    int pdoIndex, canopen_master::PDOCommunicationParameters const& parameters) {
+    pdoIndex = setupEncoderTPDOsInternal(
+        m_expected_encoder_counter_mask, 0, messages,
+        pdoIndex, parameters
+    );
```
The creation of an Internal method here does not bring anything. It was present for Analog because of the duplication between converted and raw analog readings.

doudou commented on src/DriverBase.hpp:
```diff
@@ -20,6 +20,9 @@ namespace motors_roboteq_canopen {
         base::Time m_joint_state_period;
 
 
+        uint32_t m_received_enconder_counter_mask = 0;
```
Is this compiling ? ...

doudou reviewed (commented):
You are missing the update logic in the process() call.

DeboraTahara committed:
feat: define the JointStatePositionSources enum

DeboraTahara committed:
fix: get the joint state according to the joint state position source

DeboraTahara committed:
feat: Add method to convert the encoder value to position

doudou commented on src/Factors.hpp:
```diff
@@ -39,6 +40,15 @@ namespace motors_roboteq_canopen {
         double max_current = 1000;
 
+        /** Factor to multiply the absolute encoder position value */
+        uint32_t encoder_position_factor = 1;
```
float

doudou commented on src/Factors.hpp:
```diff
@@ -39,6 +40,15 @@ namespace motors_roboteq_canopen {
+        /** Factor to multiply the absolute encoder position value */
+        uint32_t encoder_position_factor = 1;
+        /** Factor to multiply the absolute encoder speed value */
+        uint32_t encoder_speed_factor = 1;
```
float as well

doudou commented on src/JointStateSources.hpp:
```diff
@@ -0,0 +1,32 @@
+        JOINT_STATE_POSITION_SOURCE_ENCODER
+    };
+
+    /** Sources for the joint speed state */
+    enum JointStateSpeedSources {
```
Remove support for the encoder as a speed source.

doudou commented on src/Channel.cpp:
```diff
@@ -86,13 +98,13 @@ JointState Channel::getJointState() const {
         case CONTROL_SPEED:
         case CONTROL_SPEED_POSITION: {
             int32_t feedback = get<Feedback>();
-            state.speed = m_factors.relativeSpeedToSI(feedback);
+            state.speed = m_factors.speedToSI(feedback, m_joint_state_speed_source);
```
Fix that

doudou commented on src/Channel.cpp:
```diff
@@ -86,13 +98,13 @@ JointState Channel::getJointState() const {
         case CONTROL_SPEED:
         case CONTROL_SPEED_POSITION: {
             int32_t feedback = get<Feedback>();
```
The point of the source change is that while in AUTO and NONE we keep the current behaviour, in ENCODER we can read it while in OPEN_LOOP mode. Add it to the OPEN_LOOP case.

DeboraTahara committed:
feat: add unit tests for absoluteEncoderPositionToSI and positionToSI

doudou commented on src/Channel.cpp:
```diff
@@ -81,8 +81,12 @@ JointState Channel::getJointState() const {
         case CONTROL_NONE:
-        case CONTROL_OPEN_LOOP:
+        case CONTROL_OPEN_LOOP: {
+            int32_t feedback = get<EncoderCounter>();
+            state.position = m_factors.positionToSI(feedback,
+                                                    m_joint_state_position_source);
```
This is valid only if m_joint_state_position_source is encoder. Otherwise, nothing should be done at all. Add the if guard *and* call the right factor conversion method directly.

This code will break systems that don't have encoders (since get<EncoderCounter> will throw) and we would not have any way to disable it.

doudou commented on src/Channel.cpp:
```diff
@@ -92,7 +96,8 @@ JointState Channel::getJointState() const {
         case CONTROL_PROFILED_POSITION:
         case CONTROL_POSITION: {
             int32_t feedback = get<Feedback>();
```
get<Feedback> here is wrong if the source is the encoder ... the previous code was right.

doudou commented on src/Factors.hpp:
```diff
@@ -39,6 +40,11 @@ namespace motors_roboteq_canopen {
+        /** Factor to multiply the absolute encoder position value */
+        float encoder_position_factor = 1.0;
+
+        float absoluteEncoderPositionToSI(int32_t position) const;
```
`absoluteEncoderPosition` is ambiguous. An incremental encoder (what we have) gives out a position that is absolute w.r.t. an arbitrary zero (i.e. absolute with an offset). We generally do not call that "absolute position".

I think that encoderPositionToSI would be enough (or even encoderToSI)

doudou commented on test/test_Factors.cpp:
```diff
@@ -56,6 +57,19 @@ TEST_F(FactorsTest, it_returns_SI_position_max_at_relative_position_1000) {
+    ASSERT_FLOAT_EQ(0, factors.absoluteEncoderPositionToSI(0));
+}
+
+TEST_F(FactorsTest, it_returns_the_SI_position_for_absolute_encoder_minus_1000) {
+    ASSERT_FLOAT_EQ(-500, factors.absoluteEncoderPositionToSI(-1000));
+}
+
+TEST_F(FactorsTest, it_returns_the_SI_position_for_absolute_encoder_1000) {
```
Update test descriptions

doudou commented on test/test_Factors.cpp:
```diff
@@ -75,3 +89,23 @@ TEST_F(FactorsTest, it_returns_position_1000_at_SI_position_max) {
 }
+
+TEST_F(FactorsTest, it_returns_the_SI_position_for_no_joint_state_position_source) {
+    ASSERT_FLOAT_EQ(84,
+        factors.positionToSI(1000,
+            JointStatePositionSources::JOINT_STATE_POSITION_SOURCE_NONE));
```
If there is no position source ... we should definitely not return a position ....

doudou committed:
fix: disable the DS402Channel tests

They don't pass, but this part of the code is unmaintained anyways. Disable the
test for now, we'll remove the code later on

doudou committed:
chore: rename JointStateSource to match the enum name

doudou committed:
fix: finish integration on encoder as position feedback

doudou committed:
fix: base SubID for the encoders should be 1

It is the first channel's encoder (per roboteq) and is how Channel works (add
channel number - starting at zero) to the object's declared sub ID

doudou committed:
fix: size of the Feedback object

It seems that the size changed along firmware version (problem we had with other
objects)

doudou commented:
@DeboraTahara I finished the support.

WARNING: this change might very well break existing systems. Need to review the overall allocation of PDOs in the orogen component.

wvmcastro commented on test/test_Factors.cpp:
```diff
@@ -56,6 +57,19 @@ TEST_F(FactorsTest, it_returns_SI_position_max_at_relative_position_1000) {
+}
+
+TEST_F(FactorsTest, it_returns_the_SI_position_for_absolute_encoder_1000) {
+    ASSERT_FLOAT_EQ(0.5, factors.encoderToSI(1));
```
It really does not make much difference. But, the test intention is to test the limits, so why not use the limits themselves?

doudou replied:
What limits ? The encoder conversion has no limits. But need to update the test description though.

seixasxbr commented on test/CMakeLists.txt:
```diff
@@ -1,6 +1,5 @@
 rock_gtest(test_suite suite.cpp Helpers.cpp
     test_Factors.cpp
-    test_DS402Channel.cpp test_DS402Driver.cpp
```
why were these tests removed?

doudou replied:
Cf. commit message. The DS402 support never really worked and has not been used in years. It is very likely to currently be broken.

wvmcastro commented on src/Objects.hpp:
```diff
@@ -15,13 +15,14 @@ namespace motors_roboteq_canopen {
     CANOPEN_DEFINE_OBJECT(0x2100, 1, MotorAmps,                     std::int16_t);
     CANOPEN_DEFINE_OBJECT(0x2102, 1, AppliedPowerLevel,             std::int16_t);
     CANOPEN_DEFINE_OBJECT(0x210C, 1, BatteryAmps,                   std::int16_t);
+    CANOPEN_DEFINE_OBJECT(0x2104, 1, EncoderCounter,                std::int32_t);
```
I think more specific is better: `AbsoluteEncoderCounter`

doudou replied:
The fact that an incremental encoder is 'Absolute' is up to debate. We count encoder ticks from an arbitrary zero. If we miss counts, the count drifts. Thus, not Absolute

wvmcastro replied:
Ok, I was aiming for consistency with the datasheet

wvmcastro commented on src/Factors.hpp:
```diff
@@ -39,6 +40,11 @@ namespace motors_roboteq_canopen {
 
+        /** Factor to apply on encoder counts to convert to the driver's output position
+         */
+        float encoder_position_factor = 1.0;
```
is 1 a safe default? `base::unknown<float>()` would force the user a set it properly

doudou replied:
'1' says 'just show me the encoder count'. I think it's a very sane default.

wvmcastro commented on src/DriverBase.cpp:
```diff
@@ -58,10 +65,23 @@ void DriverBase::setAnalogInputEnableInTPDO(int index, bool flag) {
     }
 }
 
+void DriverBase::setEncoderCounterEnableInTPDO(int index, bool flag) {
```
the `flag` is the toggles for enable/disable?
```suggestion
void DriverBase::setEncoderCounterEnableInTPDO(int index, bool enabled) {
```

wvmcastro commented on src/DriverBase.cpp:
```diff
@@ -215,6 +239,31 @@ int DriverBase::setupAnalogTPDOs(std::vector<canbus::Message>& messages,
     return pdoIndex;
 }
 
+int DriverBase::setupEncoderTPDOs(std::vector<canbus::Message>& messages,
```
A little doc would be nice here. Each mapping must have up to 2 encoders?

doudou replied:
Check documentation of the Driver class and/or the setupEncoderTPDOs

wvmcastro commented on src/DriverBase.cpp:
```diff
@@ -41,6 +41,9 @@ canopen_master::StateMachine::Update DriverBase::process(canbus::Message const&
             m_received_converted_analog_inputs_mask |= 1 << (single_update.second - 1);
         }
+        else if (single_update.first == EncoderCounter::OBJECT_ID) {
+            m_received_encoder_counter_mask |= 1 << (single_update.second - 1);
```
where `m_received_encoder_counter_mask` is reset?

doudou replied:
Check documentation of the Driver class and/or the setupEncoderTPDOs

wvmcastro reviewed (commented):
who calls the reset methods? Like `resetEncoderCounterTracking` ?

doudou committed:
fix: Factor's test related to encoders

The values are unbounded. Fix the test to reflect that.

doudou committed:
chore: rename 'flag' to the much better 'enable' for clarity

doudou committed:
documentation

wvmcastro commented on src/Driver.hpp:
```diff
@@ -7,7 +7,86 @@
+     * The general API philosophy is to *not* deal with CAN communication. On the one
+     * hand, you need to externally instanciate a CAN driver (for instance one of the
+     * drivers/canbus package) and push packages to the driver's \c process method.
+     * On the other hand, you need
```
> On the other hand, you need

leftover or incomplete sentence
